File: src/CcDirector.Gateway.Contracts/SpokenUtterance.cs
namespace CcDirector.Gateway.Speech;

/// <summary>
/// SOMETHING THE PRODUCT IS ABOUT TO SAY OUT LOUD, WITH ITS LANGUAGE AND ITS VOICE ALREADY DECIDED
/// (issue #1031).
///
/// THE POINT OF THIS TYPE IS WHAT IT MAKES IMPOSSIBLE. Before it, a sink was handed a string, a voice and a
/// model as three separate arguments, so it was always possible - and it happened twice - to hand a sink some
/// text and no language at all. That is the hole every one of these bugs came through: the desktop resolved a
/// voice from a machine-global file and never saw the account; the browser set no language on its utterance and
/// read correct French in an English voice. Both plausible local decisions. Both silent: the audio plays.
///
/// So the language is not a parameter a caller may omit or forget. There is no parameterless construction, no
/// defaultable language, no settable-later language, and the constructor is private - the ONLY way an utterance
/// comes into existence is <see cref="For"/>, which cannot be called without naming a language. A sink whose
/// parameter is a <c>SpokenUtterance</c> cannot be passed a bare string: that is a compile error, not a test
/// failure. Issue #1008 reached for a test where a type was the stronger tool. A test tells you afterwards; a
/// type refuses.
///
/// IT CARRIES A VOICE AND IT MUST NEVER CARRY A MODEL. That is the rule the whole mission turns on. The
/// reverted build (devthrottle_internal#547) made choosing a language switch the speech MODEL, and that engine
/// could not say the lengths this product writes. A language selects the VOICE inside the one engine that
/// already serves English. There is deliberately no model, engine, endpoint or rate on this type, so there is
/// nothing here for a future change to branch an engine on - and a reflection test fails the build if a member
/// named like one appears.
///
/// WHY IT IS NOT ONE PHYSICAL SPEAKER. Some speech has to be local and network-free: when the product says it
/// cannot reach the fleet, it cannot send that sentence through the fleet to be synthesized. So there will
/// always be more than one engine - the hosted voice and the device's own - and "one place" means one CONTRACT
/// with N dumb sinks, not one speaker. This is that contract.
/// </summary>
public sealed class SpokenUtterance
{
    private SpokenUtterance(string text, SpokenLanguage language, string voice)
    {
        Text = text;
        Language = language;
        Voice = voice;
    }

    /// <summary>The words to say. Spoken CONTENT, so it carries whatever accents its language needs - and it is
    ///  therefore never written to a log or a console; log <see cref="Length"/> or the language code instead
    ///  (docs/MISSION-multilingual-RULINGS.md, guard 1).</summary>
    public string Text { get; }

    /// <summary>The language these words are in. Present on every utterance by construction.</summary>
    public SpokenLanguage Language { get; }

    /// <summary>The voice to say them with - always a voice that speaks <see cref="Language"/>, because the one
    ///  resolver that builds these decides both together.</summary>
    public string Voice { get; }

    /// <summary>How long the text is. THE log-safe fact about an utterance: a length in a log line tells you
    ///  what you need for a length bug without putting spoken content on an output channel.</summary>
    public int Length => Text.Length;

    /// <summary>
    /// Build an utterance. The only way one exists.
    ///
    /// Every argument is required, and that is the entire mechanism: there is no overload without a language,
    /// no default, and no property to set one afterwards. A caller who does not know the language cannot get
    /// past this line, which is exactly the position the old three-loose-arguments call sites were never put in.
    ///
    /// It is deliberately NOT the place that decides anything. It validates and packages what it is given; the
    /// DECISION - which language this account is spoken to in, and which voice speaks it - belongs to
    /// <c>TenantSettingsResolver</c>, which is the one caller of this method in the Gateway and is where a
    /// fourth language is added.
    /// </summary>
    /// <exception cref="ArgumentNullException">The language is null - the failure this type exists to make
    ///  loud. Never a silent English default: an account that quietly gets English is the reported bug.</exception>
    /// <exception cref="ArgumentException">The text or the voice is blank. Synthesizing nothing costs money and
    ///  returns silence, and a blank voice reaches the engine as a 422 the listener hears as silence too - so
    ///  both fail here, where the caller can see them, rather than at the far end.</exception>
    public static SpokenUtterance For(SpokenLanguage language, string voice, string text)
    {
        ArgumentNullException.ThrowIfNull(language);
        var spokenVoice = (voice ?? "").Trim();
        if (spokenVoice.Length == 0)
            throw new ArgumentException(
                "A spoken utterance needs a voice. The resolver always has one - every language has a default "
                + "voice - so a blank here means the voice was lost between the decision and this call.",
                nameof(voice));
        if (string.IsNullOrWhiteSpace(text))
            throw new ArgumentException(
                "A spoken utterance needs words. Synthesizing an empty string is billed and returns silence, "
                + "which reaches the listener as a voice that failed.", nameof(text));
        return new SpokenUtterance(text, language, spokenVoice);
    }

    /// <summary>
    /// The same utterance with different words - same language, same voice.
    ///
    /// For the one legitimate reason to change the text after the decision: a length guard that trims a runaway
    /// narration and appends the sentence telling the listener it was cut. Going back through
    /// <see cref="For"/> would mean handing a language in again at a call site that has one in its hand, which
    /// is precisely the shape this type removes.
    /// </summary>
    public SpokenUtterance WithText(string text)
    {
        if (string.IsNullOrWhiteSpace(text))
            throw new ArgumentException("A spoken utterance needs words.", nameof(text));
        return new SpokenUtterance(text, Language, Voice);
    }
}

File: packages/client-core/src/speech/spokenUtterance.ts
// SOMETHING THE BROWSER IS ABOUT TO SAY OUT LOUD, WITH ITS LANGUAGE ALREADY DECIDED (issue #1031).
//
// This is the browser expression of the SAME contract the Gateway and the desktop use - one contract, two
// runtimes, not two designs. TypeScript cannot share a C# type, so the property is rebuilt here rather than
// imported: an utterance cannot be made without a language, one factory makes them, and sinks take only that.
//
// WHY IT EXISTS. One notice in this product is spoken by the BROWSER rather than by the Gateway: the refusal
// when a menu owns a session's screen. That speech has to be local by design - it is the product saying it will
// not act, so it must cost nothing, need no network, and never be the thing that fails. It was handed a bare
// string, so it set no language, so a correctly-translated French refusal was pronounced by the device's default
// English voice. The audio played. Nothing errored. That is the failure shape this whole mission is about.
//
// IT CARRIES A LANGUAGE AND NEVER A MODEL OR AN ENGINE. A language selects the voice; a language selecting an
// engine is what got the feature reverted (devthrottle_internal#547). There is nothing on this type for a
// future change to branch an engine on - and the local engine is the device's own, which is not ours to choose.

/**
 * An utterance: words, and the language they are in. Both always present.
 *
 * `decided` is a marker, and it is load-bearing: a plain `{ text, language }` object literal does NOT satisfy
 * this type, so a sink cannot be handed something merely utterance-SHAPED that skipped the factory. It is
 * weaker than the C# side, where the constructor is private and there is no way around it - a determined caller
 * could write the marker by hand - but it turns the accident into something you have to do on purpose, and the
 * accident is what happened.
 */
export interface SpokenUtterance {
  /** The words to say. Spoken CONTENT, so it carries whatever accents its language needs - and it is therefore
   *  never logged; log its length instead (docs/MISSION-multilingual-RULINGS.md, guard 1). */
  readonly text: string;
  /** The language code the Gateway decided for this account: "en", "fr", "es". Never guessed here - the browser
   *  cannot know the account's language, and a guess here plus a guess in another shell is two different
   *  guesses. */
  readonly language: string;
  /** Set only by `utteranceFor`. See the note on this interface. */
  readonly decided: true;
}

/**
 * The languages this product speaks, as the browser knows them.
 *
 * A SECOND COPY OF A LIST, and that is a real cost stated plainly: TypeScript cannot share the C# set, so this
 * has to be kept in step by hand. It is worth it because the alternative was worse - the check here was "is the
 * string nonblank", which let "not-a-language" through to be pronounced by the device's default voice. A guard
 * test asserts this list matches the Gateway's, so drift fails the build rather than shipping.
 */
export const KNOWN_LANGUAGES: ReadonlySet<string> = new Set(["en", "fr", "es"]);

/** Whether this code names a language the product speaks. Case-insensitive; the caller trims. */
export function isKnownLanguage(code: string): boolean {
  return KNOWN_LANGUAGES.has((code ?? "").trim().toLowerCase());
}

/**
 * Build an utterance. The only way one is meant to exist.
 *
 * Both arguments are required and neither is optional, which is the mechanism: a caller who does not know the
 * language cannot get past this line. The Gateway sends the language beside the words on every response that
 * carries something to say, so a caller always has one to give.
 *
 * THROWS on a missing language rather than defaulting to English. A quiet English default is the reported bug
 * itself - the setting appears to do nothing - and every caller of this is on a path that already shows the
 * same notice on screen, so a throw is visible in the error channel instead of silently mispronouncing.
 */
export function utteranceFor(language: string, text: string): SpokenUtterance {
  const spokenLanguage = (language ?? "").trim().toLowerCase();
  if (!isKnownLanguage(spokenLanguage)) {
    throw new Error(
      `"${language}" is not a language DevThrottle speaks (known: ${[...KNOWN_LANGUAGES].join(", ")}). `
        + "NONBLANK IS NOT VALID: an unrecognized code passes a length check, then finds no device voice, and "
        + "gets spoken in the device's default - which is the silent failure this type exists to prevent.",
    );
  }
  if (typeof text !== "string" || text.trim().length === 0) {
    throw new Error("A spoken utterance needs words.");
  }
  return { text, language: spokenLanguage, decided: true };
}
